pkg/controller/gkenetworkparamset/gkenetworkparamset_controller.go

```diff
@@ -17,106 +17,176 @@ limitations under the License.
+			key, err := cache.MetaNamespaceKeyFunc(new)
+			if err == nil {
+				c.queue.Add(key)
+			}
+		},
 	})
 
-	c.informerStartFn(stopCh)
+	if !cache.WaitForNamedCacheSync("gkenetworkparamset", stopCh, c.gkeNetworkParamsInformer.HasSynced) {
+		return
+	}
+	for i := 0; i < numWorkers; i++ {
+		go wait.UntilWithContext(ctx, c.runWorker, time.Second)
+	}
 
 	<-stopCh
-	klog.Infof("done gkenetworkparamset controller")
 }
 
-func (c *Controller) onGKENetworkParamSetAdded(params *networkv1alpha1.GKENetworkParamSet) error {
-	return c.updateGKENetworkParamSet(params)
+// worker pattern adapted from https://github.com/kubernetes/client-go/blob/master/examples/workqueue/main.go
+func (c *Controller) runWorker(ctx context.Context) {
+	for c.processNextItem(ctx) {
+	}
 }
-func (c *Controller) onGKENetworkParamSetUpdated(oldParams, newParams *networkv1alpha1.GKENetworkParamSet) error {
-	return c.updateGKENetworkParamSet(newParams)
+
+func (c *Controller) processNextItem(ctx context.Context) bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+
+	defer c.queue.Done(key)
+
+	err := c.updateGKENetworkParamSet(ctx, key.(string))
+	c.handleErr(err, key)
+	return true
 }
 
-func (c *Controller) updateGKENetworkParamSet(params *networkv1alpha1.GKENetworkParamSet) (rerr error) {
-	ctx := context.Background()
-	cidrs := []string{}
+// handleErr checks if an error happened and makes sure we will retry later.
+func (c *Controller) handleErr(err error, key interface{}) {
+	if err == nil {
+		// Forget about the #AddRateLimited history of the key on every successful synchronization.
+		// This ensures that future processing of updates for this key is not delayed because of
+		// an outdated error history.
+		c.queue.Forget(key)
+		return
+	}
 
-	defer func() {
-		err := gkenetworkparamsetutil.UpdateGKENetworkParamSetStatus(ctx, c.networkClientset.NetworkingV1alpha1().GKENetworkParamSets(), params, cidrs)
-		if err != nil {
-			klog.Error(err, "failed to patch GKENetworkParamSetStatus")
-			rerr = multierr.Append(rerr, err)
-		}
-	}()
+	// This controller retries 5 times if something goes wrong. After that, it stops trying.
+	if c.queue.NumRequeues(key) < 5 {
+		klog.Warningf("Error while updating GKENetworkParamSet object, retrying %v: %v", key, err)
+
+		// Re-enqueue the key rate limited. Based on the rate limiter on the
+		// queue and the re-enqueue history, the key will be processed later again.
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	c.queue.Forget(key)
+	// Report to an external entity that, even after several retries, we could not successfully process this key
+	utilruntime.HandleError(err)
+	klog.Errorf("Dropping GKENetworkParamSet %q out of the queue: %v", key, err)
+}
+
+func (c *Controller) updateGKENetworkParamSet(ctx context.Context, key string) error {
+	obj, exists, err := c.gkeNetworkParamsInformer.GetIndexer().GetByKey(key)
+	if err != nil {
+		klog.Errorf("Fetching object with key %s from store failed with %v", key, err)
```

Comment:
special cast not found that means "it was deleted", per our conversation I think that a deletion is a no op, so we don't want to retry
```
                 // if the object was deleted don't do anythin
		if errors.IsNotFound(err) {
			return nil
		}
```

this erros come from `	"k8s.io/apimachinery/pkg/api/errors"`

Reply:
oh, you re using the index directly, I just noticed ... it may work too

Reply:
I thought this case was separated out by the exists flag. Is this not the case?

Reply:
Does this still need to be changed?

Reply:
no, I think is ok, is just that was reading with the Listers in mind